5tuple routing (#3536)
quicwg/base-drafts: draft-ietf-quic-transport.md

Servers MUST drop incoming packets under all other circumstances.

### Considerations for simple load balancers

> A server farm could be deployed behind a simple load balancer that routes

Reviewer: ```suggestion
Servers could be deployed behind a simple load balancer that routes
```

> A server farm could be deployed behind a simple load balancer that routes
> packets based on source and destination IP addresses and ports. Changes to
> the IP address or port of a peer could result in packets being routed by the load
> balancer to a different server. Such a server deployment could use one of the
> following methods to for connection continuity when a client's address changes.

Reviewer 2: ```suggestion
A server deployment could load balance among servers using simply source and
destination IP addresses and ports. Changes to the client's IP address or port of
could result in packets being forwarded to the wrong server. Such a server
deployment could use one of the following methods to for connection continuity
when a client's address changes.
```

> * Servers could use an out-of-band mechanism to forward packets or connection
> state to the correct endpoint.

Reviewer: ```suggestion
* Servers could use an out-of-band mechanism to forward packets to the correct server based
on Connection ID.
```

> * If the server deployment can use other non-load-balanced server IP
> addresses or ports than the one that the client is initiating connections on,
> servers could use the preferred_address transport parameter to request

Reviewer 2: ```suggestion
* If servers can use other dedicated server IP addresses or ports than
the one that the client is initiating connections to, they could use the
preferred_address transport parameter to request
```

> that clients move to these server addresses. Note that clients could choose
> not to use the preferred address.

Reviewer 2: ```suggestion
that clients move connections to these dedicated addresses. Note that clients could
choose not to use the preferred address.
```

> If a server behind such a load balancer does not implement a solution to
> maintain connection continuity SHOULD disallow connection migration
> via the disable_active_migration transport parameter.

Reviewer: ```suggestion
Servers behind load balancers that do not implement a solution to
maintain connection continuity SHOULD disallow connection migration
via the disable_active_migration transport parameter.
```

Author: outdated, resolving

> Server deployments that use such load balancing might still allow for

Reviewer: ```suggestion
Server deployments that use such load balancing might allow for
```

> If a server behind such a load balancer does not implement a solution to
> maintain connection continuity SHOULD disallow connection migration
> via the disable_active_migration transport parameter.
> 
> Server deployments that use such load balancing might still allow for
> creation of a stateless reset oracle. Servers should take additional
> precautions in accordance with {{reset-oracle}}.

Reviewer 2: ```suggestion
A server in a deployment that does not implement a solution to
maintain connection continuity during connection migration
SHOULD disallow it via the disable_active_migration transport
parameter.

Server deployments that use such load balancing might still allow for
creation of a stateless reset oracle, and should therefore follow the
guidelines in {{reset-oracle}}.
```

Servers MUST drop incoming packets under all other circumstances.

> ### Considerations for simple load balancers

Reviewer: ```suggestion
### Considerations for Simple Load Balancers
```

> A server deployment could load balance among servers using simply source and

Reviewer: ```suggestion
A deployment could load balance among servers using simply source and
```

> destination IP addresses and ports. Changes to the client's IP address or port of

Reviewer: ```suggestion
destination IP addresses and ports. Changes to the client's IP address or port
```

could result in packets being forwarded to the wrong server. Such a server

> deployment could use one of the following methods to for connection continuity

Reviewer: ```suggestion
deployment could use one of the following methods for connection continuity
```

when a client's address changes.

> * Servers could use an out-of-band mechanism to forward packets to the correct server based
> on Connection ID.

Reviewer 2: ```suggestion
* Servers could use an out-of-band mechanism to forward packets to the correct
server based on Connection ID.
```

* If servers can use other dedicated server IP addresses or ports than

> the one that the client is initiating connections to, they could use the

Reviewer: ```suggestion
the one that the client initially connects to, they could use the
```

> * If servers can use other dedicated server IP addresses or ports than
> the one that the client is initiating connections to, they could use the
> preferred_address transport parameter to request that clients move
> connections to these dedicated addresses. Note that clients could
> choose not to use the preferred address.

Reviewer 3: ```suggestion
* If servers can use a dedicated server IP address or port, other than the one
that the client is initially connects to, they could use the preferred_address
transport parameter to request that clients move connections to a dedicated
address. Note that clients could choose not to use the preferred address.
```

Reviewer 4: Isn't the whole point of this section that the address/port initially used by the client _isn't_ dedicated to this particular server?  It seems like the entire "other than... connects to" phrase is unnecessary.

> preferred_address transport parameter to request
> that clients move connections to these dedicated addresses. Note that clients could
> choose not to use the preferred address.

Reviewer 2: ```suggestion
preferred_address transport parameter to request that clients move
connections to these dedicated addresses. Note that clients could
choose not to use the preferred address.
```

A server in a deployment that does not implement a solution to
maintain connection continuity during connection migration

> SHOULD disallow it via the disable_active_migration transport

Reviewer 3: ```suggestion
SHOULD disallow migration using the disable_active_migration transport
```

parameter.

> Server deployments that use such load balancing might allow for
> creation of a stateless reset oracle, and should therefore follow the
> guidelines in {{reset-oracle}}.

Reviewer 3: Might as well just reiterate the existing requirement instead:

```suggestion
Server deployments that use this simple form of load balancing MUST avoid the
creation of a stateless reset oracle; see {{reset-oracle}}.
```

Fewer words, too.

Servers MUST drop incoming packets under all other circumstances.

> ### Considerations for 5-tuple routing architectures

Reviewer 2: ```suggestion
### Considerations for simple load balancers
```

> QUIC endpoints can be deployed behind a 5-tuple based routing architecture that
> delivers packets based on both the source and destination IP addresses and
> ports. When routing depends on addresses that the endpoint does not control,
> changes to the IP address or port of peers could result in packets being routed
> to a different server. The following actions could mitigate or resolve
> operational and security issues in this case:

Reviewer 2: ```suggestion
A server farm could be deployed behind a simple load balancer that routes
packets based on source and destination IP addresses and ports. Changes to
the IP address or port of a peer could result in packets being routed by the load
balancer to a different server. Such a server deployment could use one of the
following methods to for connection continuity when a client's address changes.
```

> * Endpoints can use an out-of-band mechanism to deliver packets to the correct
> destination or transfer state from the original destination.

Reviewer 2: ```suggestion
* Servers could use an out-of-band mechanism to forward packets or connection
state to the correct endpoint.
```

* A server can request that a connection be migrated to an address that is

> unique using the preferred_address transport parameter. For example, the initial

Reviewer: ```suggestion
stable using the preferred_address transport parameter. For example, the initial
```

Reviewer: If it was unique, it would be linkable, which seems sub-optimal.

Reviewer 3: Uniqueness is correct though.  The point is that the address is *in itself* sufficient for routing, for which uniqueness is necessary.  Stability is necessary for other reasons.  

That all said, it is true that linkability is a natural consequence of this.  We need to highlight that.  Servers that use this style of demultiplexing will create linkability between paths.  That's a big deal that needs to be highlighted.

Author: So is the consensus here that we want to leave it with "unique"?

Reviewer 4: It seems like the property we're intending to talk about is not uniqueness across clients, but an address that reaches that server specifically rather than a group of servers.  (Or reaches a smaller group of servers capable of internal routing based on CID, which amounts to the same thing.)  Perhaps the right way to express this is "an address that does not rely on 5-tuple routing"?

Reviewer 3: That is imprecise.  A connection ID means that you don't rely on 5-tuple routing.  The relevant point is that the destination address of packets is unique - at least for the time that the connection is active.

Author: I think we settled on "unique". resolving...

> * A server can request that a connection be migrated to an address that is
> unique using the preferred_address transport parameter. For example, the initial
> address may route to a 5-tuple based load balancer, and the preferred address
> could indicate a separate server address that does not require the use of the
> client address for routing. Note that clients could choose not to use the
> preferred address.

Reviewer 2: ```suggestion
* If the server deployment can use other non-load-balanced server IP
addresses or ports than the one that the client is initiating connections on,
servers could use the preferred_address transport parameter to request
that clients move to these server addresses. Note that clients could choose
not to use the preferred address.
```

> If a server does not implement one of the solutions above, it SHOULD send the
> disable_active_migration transport parameter to inform the client that any
> address change is likely to terminate the connection.

Reviewer 2: ```suggestion
If a server behind such a load balancer does not implement a solution to
maintain connection continuity SHOULD disallow connection migration
via the disable_active_migration transport parameter.
```

> Regardless of other mitigations, 5-tuple routing introduces new possibilities
> to create a stateless reset oracle. An attacker could tweak the source address
> or port of a packet to direct it to a different server and thus obtain the
> stateless reset token for a connection ID. Endpoints should take additional
> precautions in accordance with {{reset-oracle}}.

Reviewer 2: ```suggestion
Server deployments that use such load balancing might still allow for
creation of a stateless reset oracle. Servers should take additional
precautions in accordance with {{reset-oracle}}.
```

Note that these guarantees are the same guarantees provided for any NAT, for the
same reasons.

> ## Considerations for 5-tuple routing architectures

Author: Good idea.

QUIC servers can be deployed behind a 5-tuple based routing architecture that
delivers packets based on both the source and destination IP addresses and
ports. In such an architecture, clients that change IP address or port are
likely to be routed to a different server. There are several actions that can

> mitigate or resolve operational and security issues in this case.

Author: Is there a suggested change to this text?

* Servers can use an out-of-band mechanism to deliver packets to the correct
destination or transfer state from the original destination. Properly designed,

> this completely solves the problem and no further measures are necessary.

Reviewer 5: How would you do this without implementing a CID-based (re-) routing scheme?

Author: An example might be servers that are using QUIC-LB for just such a mapping, but they don't have a load balancer up front that is routing that way.

* Sending the disable_active_migration transport parameter informs the client
that any address change is likely to terminate the connection, which can lead it
to use more aggressive timeouts or terminate connections when its IP address

> changes.

Author: I'm not specifically referring to idle timeouts.  A client might send a PING periodically, for instance, to avoid NAT rebinding.

* The preferred_address transport parameter can provide a path that does not use

> the 5-tuple based routers.

Reviewer: So can the original path?  I suspect you have something in mind here, but I'm not completely sure what that is, so it might need to be spelled out more.

Author: OK, I'm rewriting this to be clearer.

> * Servers MUST either use different Stateless Reset Token keys, or encode the

Reviewer: I'm not a big fan of having a MUST in this section.  If we need a restriction like this, I think it should be in the section on stateless resets.

Author: I dunno, 21.9 (Stateless Reset Oracle) already breaks the seal on normative text in Security Considerations. That may or may not be bad, I'd rather that be addressed in a different PR>

client IP address and port in the Stateless Reset token. Doing neither will

> create a Reset Oracle (see {{reset-oracle}}).

Reviewer 6: The last two advices contradict against each other (or are incomplete).

The discussion about the use of preferred_address TP suggests that there is a direct path to any of the servers behind the server cluster, that are reachable from any client address. An attacker might try to use that path to let one of the servers generate a valid stateless token, and send it to the client.

As an example, consider the case of a server cluster consisting of two servers A and B. Server A is handling a connection QUIC connection that goes to client address X. If an attacker might send a packet directly to server B with source address of X, server B would send a valid stateless reset token to X, causing the connection to be reset.

Regarding specification, I think it might be better to simply refer to {{reset-oracle}} regarding the discussion of Stateless Resets, rather than trying to provide an exhaustive list of how to prevent attacks.

Author: Yes, good catch. I've rewritten it to reflect your concerns, though I shy away from just referring to the other section.

delivers packets based on both the source and destination IP addresses and
ports. In such an architecture, clients that change IP address or port are
likely to be routed to a different server. There are several actions that can
mitigate or resolve operational and security issues in this case.

* Servers can use an out-of-band mechanism to deliver packets to the correct
destination or transfer state from the original destination.

* If the server has another address where the 5-tuple based routers are not on-
path, the preferred_address transport parameter can communicate that address and
thus support changing client IP addresses without difficulty.

If a server does not implement one of the solutions above, it SHOULD send the
disable_active_migration transport parameter to inform the client that any
address change is likely to terminate the connection, which can lead it to use
strategies to avoid NAT rebinding or terminate connections when its IP address
changes.

Regardless of other mitigations, servers behind 5-tuple routing MUST do one of
the following to avoid creating a Reset Oracle ({{reset-oracle}}):

> * not send Stateless Reset under any circumstances, or

Reviewer: Isn't it valid to make the reset token a function of the CID?

Author: No, if solely a function of the CID, an attacker can rewrite the client address to route to a different server and then obtain a valid stateless reset token for use on the client.

Reviewer: Fair point, I think I was intending to generate them as a function of CID and server IP.

Author: Yes, this is similar to Kazuho's point. I'll edit.

> * use a different Stateless Reset Token key than other servers, or

Reviewer: What's a server in this context?

Author: Other members of the pool that can be routed to.

> * encode the client IP address and port in the Stateless Reset token. If using

Reviewer: I don't think you want to include port here, otherwise NAT rebinds could result in a silent blackhole.

Author: Well if the attacker tweaks the port it could go to a different server.

Reviewer: That could happen, but I expect most deployments will want to seamlessly handle NAT rebinds, so typically changing the port will not change the server.

Author: That would be wise, but this whole section is probably most important for servers in the cloud whose providers are choosing not to support QUIC-LB.

the preferred_address transport parameter, the token must also encode the

> preferred address.

Reviewer 6: @martinduke Thank you for the changes. This is better.

I would not argue against having a list, however I am still uncomfortable of this being a exhaustive list (i.e. "MUST do one of ..).

To give an example that is missing, it is safe for all the servers to share the key, assuming that they encode the server ID (that would be unique to each server) to the Stateless Reset Token.

Just giving an example of some feasible design would be much better, though by doing so, this paragraph might become a repetition of {{reset-oracle}}.

Author: Agreed I'll edit.

Note that these guarantees are the same guarantees provided for any NAT, for the
same reasons.

## Considerations for 5-tuple routing architectures

QUIC servers can be deployed behind a 5-tuple based routing architecture that
delivers packets based on both the source and destination IP addresses and
ports. In such an architecture, clients that change IP address or port are
likely to be routed to a different server. There are several actions that can
mitigate or resolve operational and security issues in this case.

* Servers can use an out-of-band mechanism to deliver packets to the correct
destination or transfer state from the original destination.

* If the server has another address where the 5-tuple based routers are not on-
path, the preferred_address transport parameter can communicate that address and

> thus support changing client IP addresses without difficulty.

Reviewer 6: The text sounds a bit too affirmative to me.

Not all clients would support the preferred_address TP, even those supporting the TP might have reasons not to migrate to the preferred address. All the server can do is *hope* that they would migrate.

Author: You're right. I forgot response to preferred_address is a SHOULD. I'll fix it tomorrow.
